core/src/main/java/tc/oc/pgm/modules/StatsMatchModule.java

```diff
@@ -0,0 +1,299 @@
+  }
+
+  @EventHandler
+  public void onPlayerDeath(MatchPlayerDeathEvent event) {
+    MatchPlayer victim = event.getVictim();
+    MatchPlayer murderer = null;
+    if (event.getKiller() != null) {
+      murderer = event.getKiller().getParty().getPlayer(event.getKiller().getId());
+    }
+    UUID victimUUID = victim.getId();
+    PlayerStats victimStats = allPlayerStats.get(victimUUID);
+
+    if (victimStats == null) {
+      allPlayerStats.put(victimUUID, new PlayerStats());
+      victimStats = allPlayerStats.get(victimUUID);
+    }
+
+    victimStats.onDeath();
+
+    PlayerStats finalVictimStats = victimStats;
+    // Without this delay the "Press LShift to dismount" message from the death minecart will
+    // override the stats message.
+    int victimTaskId =
+        event
+            .getMatch()
+            .getScheduler(MatchScope.LOADED)
+            .runTaskTimer(
+                0,
+                5,
+                () -> {
+                  NMSHacks.sendHotbarMessage(
+                      victim.getBukkit(), finalVictimStats.getBasicStatsMessage());
+                })
+            .getTaskId();
+
+    event
+        .getMatch()
+        .getScheduler(MatchScope.LOADED)
+        .runTaskLater(
+            70,
+            () -> {
+              Bukkit.getScheduler().cancelTask(victimTaskId);
+            });
+
+    if (murderer != null
+        && PlayerRelation.get(victim.getParticipantState(), murderer) != PlayerRelation.ALLY
+        && PlayerRelation.get(victim.getParticipantState(), murderer) != PlayerRelation.SELF) {
+      UUID murdererUUID = murderer.getId();
+      PlayerStats murdererStats = allPlayerStats.get(murdererUUID);
+
+      if (murdererStats == null) {
+        allPlayerStats.put(murdererUUID, new PlayerStats());
+        murdererStats = allPlayerStats.get(murdererUUID);
+      }
+
+      if (event.getDamageInfo() instanceof ProjectileInfo) {
+        murdererStats.setLongestBowKill(
+            victim
+                .getState()
+                .getLocation()
+                .distance(((ProjectileInfo) event.getDamageInfo()).getOrigin()));
+
+        murdererStats.onMurder();
+
+        MatchPlayer finalMurderer = murderer;
+        PlayerStats finalMurdererStats = murdererStats;
+        int murdererTaskId =
+            event
+                .getMatch()
+                .getScheduler(MatchScope.LOADED)
+                .runTaskTimer(
+                    0,
+                    5,
+                    () -> {
+                      NMSHacks.sendHotbarMessage(
+                          finalMurderer.getBukkit(), finalMurdererStats.getBasicStatsMessage());
+                    })
+                .getTaskId();
+
+        event
+            .getMatch()
+            .getScheduler(MatchScope.LOADED)
+            .runTaskLater(
+                70,
+                () -> {
+                  Bukkit.getScheduler().cancelTask(murdererTaskId);
+                });
+      }
+    }
+  }
```

Comment:
This method seems to be very long. Could you try to slice it into smaller ones? Would make it easier to read.

Reply:
Yeah i'm splitting it up, you will see next commit.